src/gc/collector.cpp

```diff
@@ -48,6 +39,29 @@ namespace gc {
 FILE* trace_fp;
 #endif
 
+static std::unordered_set<void*> roots;
+static std::vector<std::pair<void*, void*>> potential_root_ranges;
+
+// BoxedClasses in the program that are still needed.
+static std::unordered_set<BoxedClass*> class_objects;
+
+static std::unordered_set<void*> nonheap_roots;
```

Comment:
Yeah. For the class objects, it's possible that this whole class-freeing problem goes away with a moving collector (which won't have a sweep phase in which to look at classes) and in most programs the sets won't contain more than a few hundred objects so I'll put that off for now.